File: py/ventora_billing/src/ventora_billing/plans.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class BillingPlan:
    id: str
    name: str
    features: list[str]
    prices: dict[str, str]  # {"month": "price_xxx", "year": "price_yyy"}
    description: str | None = None
    limits: dict[str, Any] | None = None
    trial_days: int | None = None
    is_default: bool = False


class PlanRegistry:
    def __init__(self, plans: list[BillingPlan]) -> None:
        if not plans:
            raise ValueError("PlanRegistry requires at least one plan")
        ids = [p.id for p in plans]
        if len(ids) != len(set(ids)):
            raise ValueError("Duplicate plan IDs")
        all_prices = [pid for p in plans for pid in p.prices.values()]
        if len(all_prices) != len(set(all_prices)):
            raise ValueError("Duplicate price IDs across plans")
        self._plans = plans
        self._by_price: dict[str, BillingPlan] = {
            pid: plan for plan in plans for pid in plan.prices.values()
        }
        self._by_id: dict[str, BillingPlan] = {p.id: p for p in plans}

    @property
    def plans(self) -> list[BillingPlan]:
        return list(self._plans)

    def resolve_plan_from_price_id(self, price_id: str | None) -> BillingPlan | None:
        if not price_id:
            return None
        return self._by_price.get(price_id)

    def get_price_id(self, plan_id: str, cadence: str = "month") -> str:
        plan = self._by_id.get(plan_id)
        if not plan:
            raise KeyError(f"Plan not found: {plan_id!r}")
        price = plan.prices.get(cadence)
        if not price:
            raise KeyError(f"No price for plan {plan_id!r} with cadence {cadence!r}")
        return price

    def has_feature_access(self, plan_id: str, feature: str) -> bool:
        plan = self._by_id.get(plan_id)
        if not plan:
            return False
        return feature in plan.features

    def get_default_plan(self) -> BillingPlan | None:
        return next((p for p in self._plans if p.is_default), None)

    def get_trial_days(self, plan_id: str) -> int:
        plan = self._by_id.get(plan_id)
        if not plan or plan.trial_days is None:
            return 30
        return plan.trial_days


def create_plan_registry(plans: list[BillingPlan]) -> PlanRegistry:
    return PlanRegistry(plans)

File: packages/billing/src/plans.ts
export type BillingCadence = "month" | "year";

export type BillingPlan = {
  id: string;
  name: string;
  description?: string;
  features: string[];
  limits?: Record<string, number | string | null>;
  prices: Partial<Record<BillingCadence, string>>; // Stripe price IDs
  trialDays?: number;
  isDefault?: boolean;
};

export interface PlanRegistry {
  readonly plans: readonly BillingPlan[];
  resolvePlanFromPriceId(priceId: string | null | undefined): BillingPlan | null;
  getPriceId(planId: string, cadence?: BillingCadence): string; // throws if not found
  hasFeatureAccess(planId: string, feature: string): boolean;
  getDefaultPlan(): BillingPlan | null;
  getTrialDays(planId: string): number; // returns plan.trialDays ?? 30 as default
}

export function createPlanRegistry(plans: BillingPlan[]): PlanRegistry {
  if (plans.length === 0) {
    throw new Error("PlanRegistry requires at least one plan");
  }

  const idSet = new Set<string>();
  for (const plan of plans) {
    if (idSet.has(plan.id)) {
      throw new Error(`Duplicate plan ID: ${plan.id}`);
    }
    idSet.add(plan.id);
  }

  const priceIdMap = new Map<string, BillingPlan>();
  for (const plan of plans) {
    for (const priceId of Object.values(plan.prices)) {
      if (priceId !== undefined) {
        if (priceIdMap.has(priceId)) {
          throw new Error(`Duplicate price ID: ${priceId}`);
        }
        priceIdMap.set(priceId, plan);
      }
    }
  }

  const planMap = new Map<string, BillingPlan>(plans.map((p) => [p.id, p]));

  return {
    get plans(): readonly BillingPlan[] {
      return plans;
    },

    resolvePlanFromPriceId(priceId: string | null | undefined): BillingPlan | null {
      if (priceId == null) return null;
      return priceIdMap.get(priceId) ?? null;
    },

    getPriceId(planId: string, cadence: BillingCadence = "month"): string {
      const plan = planMap.get(planId);
      if (plan === undefined) {
        throw new Error(`Plan not found: ${planId}`);
      }
      const priceId = plan.prices[cadence];
      if (priceId === undefined) {
        throw new Error(`No price found for plan "${planId}" with cadence "${cadence}"`);
      }
      return priceId;
    },

    hasFeatureAccess(planId: string, feature: string): boolean {
      const plan = planMap.get(planId);
      if (plan === undefined) return false;
      return plan.features.includes(feature);
    },

    getDefaultPlan(): BillingPlan | null {
      return plans.find((p) => p.isDefault === true) ?? null;
    },

    getTrialDays(planId: string): number {
      const plan = planMap.get(planId);
      if (plan === undefined) return 30;
      return plan.trialDays ?? 30;
    },
  };
}
